### PR TITLE
Feature/ui

### PR DESCRIPTION
**Pull request by angelRev**
Adds mindmeld UI. Check README for how to test it.

### CODE REVIEW

**Comment by jjjacksn** [comment]
Given that the new code is nodejs (not python) have we considered putting this in a separate repo? What are the pros and cons?

File: mindmeld-ui/README.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+npm start
+```
+
+### 3. Setup configurations (Optional)
+
+At the root directory, you can add a `.env` file that includes your configurations. This file should look like this:
+
+```
+REACT_APP_GOOGLE_KEY = <YOUR_GOOGLE_KEY> # Set this if you want to use Google Cloud's Speech API (as opposed to Chromes native ASR).
+REACT_APP_MAX_ALTERNATIVES = <MAX_ALYTERNATIVES> # Set this if you want to control how many responses you get from the ASR.
```
Comment (jjjacksn) [inline mindmeld-ui/README.md]: Typo: ~~MAX_ALYTERNATIVES~~ -> MAX_ALTERNATIVES

File: mindmeld-ui/README.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+REACT_APP_MINDMELD_URL = <MINDMELD_URL> # Set this if you need to point the UI to point the UI to a URL different than the default 'http://127.0.0.1:7150/parse'
```
Comment (jjjacksn) [inline mindmeld-ui/README.md]: I suggest that we use a better namespace than `REACT_APP` for these env vars. Maybe `MINDMELD_UI` and `MINDMELD_UI_NLP_URL` for REACT_APP_MINDMELD_URL`

File: mindmeld-ui/src/app/actions.js [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+export const resetState = () => {
+  return {
+    type: APP.RESET_STATE
+  };
+};
+
+export const startRecognition = (handler) => {
+    return {
+        type: APP.START_RECOGNITION,
+        handler,
```
Comment (jjjacksn) [inline mindmeld-ui/src/app/actions.js]: Is this a function? I think it's an anti pattern to include functions in action objects (except in support of thunks or async actions, or something else processed by a store middleware).

File: mindmeld-ui/README.md (hunk 2 above)
Comment (angelRev) [reply mindmeld-ui/README.md]: React requires to prefix env variables with `REACT_APP_` [here](https://facebook.github.io/create-react-app/docs/adding-custom-environment-variables#adding-development-environment-variables-in-env). We can change that, but we would need to `eject`, and that's something I'm trying to avoid.

**Comment by minhtuev** [comment]
@jjjacksn : I think for the time being we can keep the UI part of MindMeld here for ease of development - the developer can check out the project in one place. When the UI feature becomes sufficiently complicated we might decide to move it to its own separate repo.

File: mindmeld-ui/src/app/middleware.js [hunk 4]
```diff
@@ -53,7 +50,7 @@ export default store => next => async action => {
         break;
     }
 
-    default: {
+      default: {
```
Comment (chadoakley) [inline mindmeld-ui/src/app/middleware.js]: Was this intentional?

File: source/mindmeld_ui/mindmeld_ui.rst [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+To run the app simply do:
+
+.. code-block:: console
+
+    npm start
+
+3. Setup configurations (Optional)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+At the `mindmeld-ui` directory, you can add a `.env` file that includes your configurations. This file should look like this:
```
Comment (chadoakley) [inline source/mindmeld_ui/mindmeld_ui.rst]: "In the `mindmeld-ui` directory, you can add an optional `.env` file with extended configuration. This file can include the following options:"

You could also include an example .env file in the repo called `.env-example` or something similar.

File: mindmeld-ui/src/app/middleware.js (hunk 4 above)
Comment (angelRev) [reply mindmeld-ui/src/app/middleware.js]: Fixed

File: source/mindmeld_ui/mindmeld_ui.rst (hunk 5 above)
Comment (angelRev) [reply source/mindmeld_ui/mindmeld_ui.rst]: Added.

File: mindmeld-ui/src/app/actions.js (hunk 3 above)
Comment (angelRev) [reply mindmeld-ui/src/app/actions.js]: Fixed

File: mindmeld-ui/README.md (hunk 1 above)
Comment (angelRev) [reply mindmeld-ui/README.md]: Fixed

File: mindmeld-ui/package-lock.json [hunk 6]
```diff
@@ -0,0 +1,13099 @@
+{
```
Comment (minhtuev) [inline mindmeld-ui/package-lock.json]: do we need to check in this file?

Comment (chadoakley) [reply mindmeld-ui/package-lock.json]: Yup!
https://docs.npmjs.com/files/package-lock.json
Here's the writeup on what that file is for if you're curious.

File: mindmeld-ui/src/recognition/languages.js [hunk 7]
```diff
@@ -0,0 +1,482 @@
+export default [
+  /*
```
Comment (minhtuev) [inline mindmeld-ui/src/recognition/languages.js]: Should we uncomment these languages?

Comment (minhtuev) [reply mindmeld-ui/src/recognition/languages.js]: Or maybe we can just remove them.

File: mindmeld-ui/yarn.lock [hunk 8]
```diff
@@ -0,0 +1,8714 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Comment (minhtuev) [inline mindmeld-ui/yarn.lock]: Do we need to keep this file?

File: source/mindmeld_ui/mindmeld_ui.rst [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In order to run the app, you need to install it's dependencies. You can simply run:
+
+.. code-block:: console
+
+    npm install
+
+
+2. Run the app
```
Comment (minhtuev) [inline source/mindmeld_ui/mindmeld_ui.rst]: Run the web app

File: source/mindmeld_ui/mindmeld_ui.rst [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+3. Setup configurations (Optional)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+At the `mindmeld-ui` directory, you can add a `.env` file that includes your configurations. This file should look like this:
+
+.. code-block:: console
+
+    REACT_APP_GOOGLE_KEY = <YOUR_GOOGLE_KEY> # Set this if you want to use Google Cloud's Speech API (as opposed to Chromes native ASR).
+    REACT_APP_MAX_ALTERNATIVES = <MAX_ALTERNATIVES> # Set this if you want to control how many responses you get from the ASR.
+    REACT_APP_MINDMELD_URL = <MINDMELD_URL> # Set this if you need to point the UI to point the UI to a URL different than the default 'http://127.0.0.1:7150/parse'
```
Comment (minhtuev) [inline source/mindmeld_ui/mindmeld_ui.rst]: Maybe you can add a short section:

```
4. Navigate the app
First, you will need to run the Mindmeld app. If you are developing locally, you can do:

`python -m [app-name] run`

[include the diagram]
On the left, you can select a query from a predefined list of queries from our sample blueprints. You can type the query directly into the input box. You can also click on the microphone button for sound input using automatic speech recognition.

After the query is sent to the MindMeld application, the Web UI receives a response and displays each component visually. You can navigate to each component (such as Intent Classifier) to learn more about them.
```

Probably we will add more documentation for the UI but we can do that in a separate PR.

File: mindmeld-ui/src/recognition/languages.js (hunk 7 above)
Comment (angelRev) [reply mindmeld-ui/src/recognition/languages.js]: Removed the whole file, turns out we were not using it!

File: source/mindmeld_ui/mindmeld_ui.rst (hunk 9 above)
Comment (angelRev) [reply source/mindmeld_ui/mindmeld_ui.rst]: Fixed

**Comment by minhtuev** [comment]
Don't worry about the Travis CI build failure - it seems that some LSTM tests are running longer than expected. I will fix that in another PR.

File: mindmeld-ui/yarn.lock (hunk 8 above)
Comment (angelRev) [reply mindmeld-ui/yarn.lock]: Yes, this is so people can use either `npm` or `yarn` to manage the dependencies.

File: source/mindmeld_ui/mindmeld_ui.rst (hunk 10 above)
Comment (angelRev) [reply source/mindmeld_ui/mindmeld_ui.rst]: Added!
